File: leetcode/166-fraction-to-recurring-decimal.py
class Solution(object):
    def fractionToDecimal(self, numerator, denominator):
        """
        :type numerator: int
        :type denominator: int
        :rtype: str
        """
        num, den = numerator, denominator
        symb = '-' if ((num>>31)^(den>>31)) and num else ''

        num, den = abs(num), abs(den)

        rem = num % den
        num = int(num / den)
        inte = str(num)

        frac = ''
        rems = {}
        p = 0
        while rem and rem not in rems:
            rems[rem] = p
            num = rem * 10
            rem = (num % den)
            num = int(num / den)
            p += 1
            frac += str(num)

        if len(rems) > 0:
            if rem > 0:
                p = rems[rem]
                frac = frac[:p] + '(' + frac[p:] + ')'
            frac = '.' + frac
        return symb + inte + frac

File: leetcode/166-fraction-to-recurring-decimal.cpp
class Solution
{
 public:
  string fractionToDecimal(int numerator, int denominator)
  {
    long long num = numerator, den = denominator;
    string symb = (((num >> 31) ^ (den >> 31)) && num) ? "-" : "";
    num = abs(num);
    den = abs(den);

    long long rem = num % den;
    num /= den;
    string inte = to_string(num);

    string frac;
    map<long long, int> rems;
    for (int i = 0; rem && rems.find(rem) == rems.end(); ++i) {
      rems[rem] = i;
      num = rem * 10;
      rem = num % den;
      num /= den;
      frac += to_string(num);
    }

    frac = frac.empty()
           ? "" : "." +
           (!rem
            ? frac
            : (frac.substr(0, rems[rem]) +
               "(" + frac.substr(rems[rem]) + ")"));
    return symb + inte + frac;
  }
};
